src/main/java/com/fzakaria/mvn2nix/maven/Maven.java

```diff
@@ -0,0 +1,126 @@
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.stream.Collectors;
+import java.util.stream.StreamSupport;
+
+/**
+ * Small utility wrapper around maven-invoker.
+ * https://maven.apache.org/shared/maven-invoker/index.html
+ */
+public class Maven {
+
+    private final Invoker invoker;
+    private final File localRepository;
+
+    /**
+     * Constructor.
+     * The local repository must be a directory.
+     * @param localRepository The location of the local repository
+     */
+    public Maven(File localRepository) {
+        Preconditions.checkArgument(localRepository.isDirectory());
+        this.localRepository = localRepository;
+        this.invoker = new DefaultInvoker();
+        this.invoker.setLocalRepositoryDirectory(localRepository);
+        this.invoker.setLogger(new Log4j2Logger());
+
+        // send all of maven's output to log4j2 which goes to STDERR
+        PrintStreamHandler handler = new PrintStreamHandler(
+                IoBuilder.forLogger("maven-invoker").setAutoFlush(true).buildPrintStream(),
+                true
+        );
+
+        this.invoker.setOutputHandler(handler);
+        this.invoker.setErrorHandler(handler);
+    }
+
+    /**
+     * Create a {@link Maven} object with a temporary local repository.
+     * The local repository is deleted upon JVM exit.
+     */
+    public static Maven withTemporaryLocalRepository() {
+        try {
+            File tempLocalRepository = Files.createTempDirectory("mvn2nix").toFile();
+            tempLocalRepository.deleteOnExit();
+            return new Maven(tempLocalRepository);
+        } catch (IOException e) {
+            throw new UncheckedIOException(e);
+        }
+    }
+
+    public void executeGoals(String... goals) throws MavenInvocationException {
+        InvocationRequest request = new DefaultInvocationRequest();
+        request.setGoals(Lists.newArrayList(goals));
+        request.setBatchMode(true);
+
+        InvocationResult result = this.invoker.execute(request);
+        if (result.getExitCode() != 0) {
+            throw new MavenInvocationException(
+                    String.format("Failed to execute goals [%s]. Exit code: %s", Arrays.toString(goals), result.getExitCode()),
+                    result.getExecutionException()
+            );
+        }
+    }
+
+    /**
+     * Walk the local repository and collect all the artifacts downloaded
+     * https://cwiki.apache.org/confluence/display/MAVEN/Remote+repository+layout#:~:text=Repository%20metadata%20layout-,Introduction,versioned%20releases%20of%20dependent%20projects
+     * The layout is assumed to be:
+     * ${groupId.replace('.','/')}/${artifactId}/${version}/${artifactId}-${version}${classifier==null?'':'-'+classifier}.${type}
+     * @return
+     */
+    public Collection<Artifact> collectAllArtifactsInLocalRepository() {
```

Comment:
TODO: Determine if we should support _platformId_
https://cwiki.apache.org/confluence/display/MAVEN/Remote+repository+layout#:~:text=Repository%20metadata%20layout-,Introduction,versioned%20releases%20of%20dependent%20projects